### ISSUE TITLE
[next] <py-config> doesn't handle syntax errors well

### ISSUE THREAD

**Issue by antocuni** [pyscript/pyscript#1700]
I have noticed that when <py-config> contains invalid TOML, we have several unexpected things, but I could not find a common pattern so I'm reporting all of them here.

# Case 1: invalid syntax
```html
<!doctype html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>PyScript Error</title>
        <script type="module" src="https://cdn.jsdelivr.net/npm/@pyscript/core@latest"></script>
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@pyscript/core/dist/core.css">
        <py-config>
          [[fetch]]
              files = ["pyterminal.py"]this is a syntax error
        </py-config>

        <script type="py">
          print('hello')
        </script>
    </head>
</html>
```

Here I correctly get an error in the console:
![image](https://github.com/pyscript/pyscript/assets/344507/2d53a9f6-a7ae-4fcf-b0f3-04d9a73ef44e)

However:
1) errors should be displayed to the DOM by default
2) the error message says something about JSON which is very confusing, since `<py-config>` is supposed to contain TOML

# Case 2: another invalid syntax, completely ignored
```html
        <py-config>
          this is invalid syntax
        </py-config>

        <script type="py">
          print('hello')
        </script>
```

In this case, the `py-config` is just silently ignored and `hello` is normally printed.
I would expect that in case of syntax error in the config, the execution of the page stops and the error is displayed to the DOM.

# Case 3: I cannot really understand what's happening
```html
        <py-config>
          [[fetch]]
          files = 
        </py-config>

        <script type="py">
          print('hello')
        </script>
```
Here I get *yet another* error, but also I see that it tries to fetch `config.txt`, which I don't really know where it comes from :thinking: 

![image](https://github.com/pyscript/pyscript/assets/344507/20ee48e6-e918-4ca9-bf32-e86e6f10fdeb)

**Comment by ntoll** [comment]
Please bear in mind that the TOML is parsed to JSON (hence the JSON related errors). I believe we should be rather unforgiving when parsing TOML->JSON and just fail at all costs with something like "Invalid TOML in configuration".

**Comment by fpliger** [comment]
What @ntoll said :)

**Comment by WebReflection** [comment]
we don't have a full blown TOML parser here so if you want that, consider adding many extra KB to the core package, otherwise I think the issue is always related to the config which, if invalid TOML or JSON, should be handled equally well in error reporting. This is also like a duplicate bug for this other issue https://github.com/pyscript/pyscript/issues/1699 as in there the `worker` attribute has nothing to do with TOML or config causing errors.

**Comment by WebReflection** [comment]
P.S. the TOML parser is lazy, JSON costs nothing to core or PyScript / polyscript users, so it's really up to you telling me how much a TOML full parser should delay PyScript bootstrap for the few fields we handle in our config files. I am OK either ways, but so far nobody had issues with code that works, I am not sure we want to cover all possible mistakes anyone can do anywhere but surely we want to inform users better than we do now.

**Comment by antocuni** [comment]
> Please bear in mind that the TOML is parsed to JSON (hence the JSON related errors). I believe we should be rather unforgiving when parsing TOML->JSON and just fail at all costs with something like "Invalid TOML in configuration".

+1 for that.
It's ok to have our custom TOML parser if it saves from kb, but it should always reject invalid TOML.
If it accepts invalid TOML, we should either fix it or remove it, IMHO.

And yes, errors should be reported clearly.

**Comment by WebReflection** [comment]
> it should always reject invalid TOML

that requires a 100% TOML specs compliant parser, hence bloat, for those max 8 (???) fields of TOML we have ... we need to make a conscious choice here: bloat the whole project which was promoted as small + faster or bring in a TOML parser.

Happy to hear/explore alternatives though.

**Comment by WebReflection** [comment]
Note a part:

This MR is a blocker for this one https://github.com/pyscript/pyscript/pull/1698

It needs to go in otherwise we don't have the error plugin when the config error eventually happens (or the bootstrap fails)

**Comment by antocuni** [comment]
> > it should always reject invalid TOML
> 
> that requires a 100% TOML specs compliant parser,

from my point of view, "slow but correct" is better than "fast but incorrect".
But I am not well versed enough in web things to correctly estimate what is the real impact of a full TOML parser on any actual example.

E.g.:
- for a simple `<script type="py">` pyodide hello world: what is the % of size increase and the measured extra time?
- same question but for micropython.

Happy to explore alternatives of course, and curious to hear what others think.

**Comment by ntoll** [comment]
Warning: very opinionated opinion coming.... don't judge (but I welcome constructive feedback and discussion - which I know I'll get from you folks anyway) 😉 

We're in the world of the web. Let's just use/promote JSON for config, and explain that as a courtesy to the Python world we make a "best effort" attempt to parse correct TOML - but there be dragons. 🐉 

Size *does* matter in the world of the web (especially on mobile). Including a whole TOML parser is a fly in the ointment for end users who expect snappy startup time (i.e. if we include a TOML parser, we're blocked from doing anything until said parser has been downloaded in order to parse the config).

"We recommend using JSON" = JSON is already built in, and since we're in the context of a browser, JSON shouldn't feel an odd choice. Furthermore, I like the "but as a courtesy to Python we make a best attempt at TOML" because it promotes choice without too much cost.

</end opinionated opinions>

😛

**Comment by WebReflection** [comment]
@ntoll basically you wrote the reason I've created [this module](https://github.com/WebReflection/basic-toml#readme) as there was nothing that small / insignificant out there ... what I think we should do, and probably we were doing it properly before, is the following (and this might be a *polyscript* responsibility as that's where the config parsing happens):

  * we use the tiny *basic-toml* module and try/catch its ability to parse
  * if there are errors, we inform that the TOML is malformed (right now I think I am not expecting that)
  * if there are no errors, we move into the next validation step, which would **the same** for JSON config
  * **validation steps**
    * we check that all fields we handle, if present, represent valid data ... that is an `Array` for packages or plugins, strings or boolean for other things, following all things we documented already
    * if any of these fields doesn't match type or values expectations (i.e. empty packages or files should throw right away) we inform the user the config is not valid

These steps should cover all possible cases but there's more to it, I believe, as also files that can't be moved in the virtual FS should throw for a reason.

In short, the whole config story must be improved in terms of errors, right now I haven't considered people writing bad configs as I consider myself that to be a very rare case but it's true that if such case happens we better provide a meaningful error.

Thoughts?

**Comment by antocuni** [comment]
> We're in the world of the web. Let's just use/promote JSON for config,

We have already discussed this [in a very long discussion](https://github.com/pyscript/pyscript/issues/701). There were pros and cons for both options. "JSON is web native" and "TOML is bigger" were both points that we have already considered, and we decided to use TOML anyway.

I'm not saying that we cannot revisit the decision, but in that case we should do it explicitly, re-evaluate the alternatives, and make sure that all the people involved in the previous discussion acknowledge the change. Else it's just disrespectful to the time and effort that the people spent in doing the previous discussion, IMHO.

> Size does matter in the world of the web (especially on mobile). Including a whole TOML parser is a fly in the ointment for end users who expect snappy startup time (i.e. if we include a TOML parser, we're blocked from doing anything until said parser has been downloaded in order to parse the config).

I'm happy to buy this argument only if we can quantify the size and time penalty, else it's just premature optimization, again IMHO.

**Comment by antocuni** [comment]
> I haven't considered people writing bad configs as I consider myself that to be a very rare case

I think it's much more frequent that you think :).
I started this discussion because *I* wrote an invalid config and took 10 minutes to understand what was going on 😂

**Comment by ntoll** [comment]
@antocuni to be clear - I wasn't meaning to be disrespectful. 

However, I think we're in a very different place to when we last had the TOML/JSON discussion WRT config. I certainly support your point about discussing things... but bear in mind that the solution we have already supports TOML if only to a limited extent, and there is an undeniable performance advantage to using JSON (it's already built in!) over TOML. In fact, if memory serves me correctly, the "next" effort for PyScript arose because @WebReflection was asked to look at the "classic" codebase with a view to making it smaller, more efficient and idiomatically web-ish so we started up quickly. A big factor in the slowness was that we had a full TOML parser to include.

To clarify my position - of course we should support TOML, that's what we decided... and I agree with you that it would be disrespectful not to (hence my "as a courtesy to Pythonistas" point). But we also support JSON and (here's the change in my position), the emphasis / recommendation should be to use JSON while acknowledging one could use TOML (that we parse on a best effort basis).

Ultimately, we need a focussed and friendly discussion to ensure we're aligned with what we're actually offering. I'm happy to facilitate this (I feel a Zoom call, or a section of PyScript FUN should be dedicated to this). The most important thing for me is that I trust you all - despite the various technical differences I know we have, I know we'll all engage with the very best of intentions as we try to explore this puzzle. :+1:

**Comment by ntoll** [comment]
>>    I haven't considered people writing bad configs as I consider myself that to be a very rare case
>
> I think it's much more frequent that you think :).
> I started this discussion because I wrote an invalid config and took 10 minutes to understand what was going on 😂

Exactly. Hence [my suggestion](https://github.com/pyscript/pyscript/issues/1700#issuecomment-1717979871) that we're unforgiving of bad TOML and generous in our error message (to which @fpliger :+1:).

What's that saying..? Just when you think you've covered all the possible ways that idiots could break your software, the universe creates a new and yet more creative set of idiots? (I include myself in that set)

:rofl:

**Comment by WebReflection** [comment]
> was asked to look at the "classic" codebase with a view to making it smaller, more efficient and idiomatically web-ish so we started up quickly.

I think there are two major topics apparently very hard to let it go for *Next* but that we should *promote* as best practices with native performance (Web or parsing) win:

  * promote the usage of `<script type="py">` over `<py-script>`
  * explain that *JSON* configs are the fastest option over *TOML*

If we go out like nothing changed and nothing should change it'll be PyScript Classic II soon :sweat_smile: and I hope we can do better to promote the *migration* pattern to better suit our goals and purposes.

That being said, I don't have any strong opinion about *JSON* vs *TOML* but I do think having an entire *TOML* parser just to parse zero to 8 fields is the opposite of premature optimization, is overkill and over engineering at its best.

I'd rather focus on a *parsing config algorithm* as explained [in here](https://github.com/pyscript/pyscript/issues/1700#issuecomment-1719103005) because even if it's not rare to have badly written TOML, all Antonio witnessed is the lack of clear error message saying: "*sorry dude, the config is a bit bananas*".

To me that use case doesn't need a fully standard *TOML* parser, and neither does [our config](https://docs.pyscript.net/latest/reference/elements/py-config.html).

**Comment by ntoll** [comment]
> "sorry dude, the config is a bit bananas"

Can we have `"Config is 🍌 (check your TOML)"` as the error message? </joke>

**Comment by WebReflection** [comment]
@antocuni @ntoll @fpliger OK folks, beside the lovely conversation, as I am working on this issue (as it is a blocker for release to me) I'd like to expose my findings:

  * we do have meaningful and documented errors in PyScript ... we probably want to be consistent with those too and `ErrorCode.BAD_CONFIG` is a great example here
  * my biggest concern is about somehow "*ruining*" *polyscript* with unnecessary bloat. We have already various users creating amazing things and they love it the way it is and nobody complained about the *config* there ... however ...
  * inPyScript we're currently borrowing the *lazy basic-toml* parser, but nothing really stops us to add in PyScript a *better TOML parser* than what *polyscript* offers, as that's lazy anyway, so we don't overlap or bloat more than necessary anything in doing so
  * I find the previous point a bit of a must because *basic-toml* is indeed ... well ... **too basic** and it assumes correct TOML but it accepts even a TOML with `shenaingans()` in it without throwing anything

As summary, I am more convinced now that I've put my hands on the code and tested various scenarios that we indeed *want*, and likely *need*, to have better TOML validation, otherwise my basic parser will silently return an empty config which would be extremely misleading to me.

None of this side-effects *polyscript*, but *PyScript* users (including ourselves) will surely benefit from such "*better TOML story*".

I would love to hear any recommendation for a better TOML parser and it's worth saying that if we don't want to make its bootstrap lazy we can just embed that code within PyScript Next as we've done already to date, hopefully not bloating too much the final result in terms of size after compression (but I will provide data around it and we'll see if we're happy).

What do you all think?

**Comment by ntoll** [comment]
OK... so there are two aspects to this... and I think we're getting to a good endpoint.

1. Not bloating PolyScript.
2. Recommended practices.

Regarding 1... I like your idea @WebReflection ~ by including a full TOML parser in PyScript (not PolyScript) we get what we want. I'd add that we should remove the TOML parsing from PolyScript and just get the TOML parser in PyScript to emit the JS object we'd get from pure JSON in any case (which is, I believe how it works right now). I'm still not convinced we need a full TOML though... I remain unconvinced we need it if we follow my suggested unforgiving but explicit mode of feedback.

As far as 2 is concerned I'm very :+1 to this (since I'm writing the docs right now):

> * promote the usage of <script type="py"> over <py-script>
> * explain that JSON configs are the fastest option over TOML

Although we should clearly document there's a choice (and the alternative of using `<py-script>` and TOML). Let many flowers bloom. :wink:

**Comment by WebReflection** [comment]
Polyscript has no TOML parsing ... or better, it uses the smallest parser out there lazily, only if the config is TOML, so it doesn't affect PyScript by any mean ... **but** I've found the "official" [toml library](https://github.com/iarna/iarna-toml#tomlparsestr--object-example) which has a relatively small export and it produces a very nice error ... it adds, once minified and brotlied, just ~8K to the project (if embedded) but it could be resolved dinamically via jsdelivr ... my biggest concern around that module is that **it's based on `eval` and it stinks** because it does a `try/catch` to understand if it's running in NodeJS or not.

This is a bummer ... but I keep investigating if there is a better option there without eval and without the bloat.

**The current issue** is that we never show errors unless the config is really badly broken ... I might be OK with that, but as the TOML parser is loaded lazily **anyway** I think we can use one slightly bigger but more correct.

**Comment by WebReflection** [comment]
FYI this is the CDN URL and it loads pretty fast ... we resolve the jsdelivr CDN anyway so it's not a big deal but that `eval("require(...)")` within its code **is** a big deal because any CSP based site would avoid `eval` from foreign modules https://cdn.jsdelivr.net/npm/@iarna/toml/parse-string/+esm

**Comment by WebReflection** [comment]
Sorry for late answer about this:

> get the TOML parser in PyScript to emit the JS object we'd get from pure JSON in any case (which is, I believe how it works right now)

yes, the config is inevitably an object ... if it comes from JSON or TOML it doesn't matter, and every TOML parser returns a JSON friendly object literal ... maybe I can change Polyscript to work with JSON or objects literal only and accept a foreign callback to parse anything else otherwise (yaml, toml, you name it) but the issue is that we cannot pass callbacks to workers but workers need to understand config files / objects ... although we can pre-parse these *before* we create a worker so everything is possible, but that'd be a breaking change for polyscript ... and I thought we were in "*feature freeze*" for this release ... I say we get main PyScript config story right (right lib, right logic) and **then** we move forward fine-tuning everything else **after** the release. From the documentation point of view, as we're releasing an alpha/beta/release candidate/whatever, I think we should just flag *worker* experimental, or in the making, or not fully available, because we have other issues to solve for workers in PyScript, not just the TOML story (we don't have plugins there, there is an issue already for that and it's for the post-release time).

**Comment by WebReflection** [comment]
An old gotcha stuck in year 2022 apparently ... https://github.com/iarna/iarna-toml/issues/45#issuecomment-1719523055 I hope they'll consider my suggestion there.

**Comment by antocuni** [comment]
wow, tons of discussion to answer to. Let me try to answer to some of the points, I hope I don't forget anything important.

---

re @ntoll 

> @antocuni to be clear - I wasn't meaning to be disrespectful.

sorry @ntoll, maybe I have been a bit too harsh. I didn't mean that you are disrespectful (you never are :green_heart:), but that if we change a big design decision "in a corner" without further discussion, *that* would be disrespectful.

> To clarify my position - of course we should support TOML, [cut] (hence my "as a courtesy to Pythonistas" point). 
> [cut] the emphasis / recommendation should be to use JSON while acknowledging one could use TOML (that we parse on a best effort basis).

to clarify my position:
- personally, I don't promote TOML because it's a "courtesy to pythonistas". I don't care about that point. I promote TOML because I think it has many advantages compared to JSON, especially for beginners (e.g., comments, and in general a better readability).

- I don't think we should do a "best effort". If we declare that we support TOML, we should support TOML. If we decide to support a half-baked custom format which resembles TOML, please don't call it TOML, else it just add confusion.

---

re @WebReflection 

> in PyScript we're currently borrowing the lazy basic-toml parser, but nothing really stops us to add in PyScript a better TOML parser than what polyscript offers, as that's lazy anyway, so we don't overlap or bloat more than necessary anything in doing so

yes, +1 for this. It sounds like the best compromise:
- advanced polyscript users can have a fast & small core
- "basic pyscript" users have a "better" and easier to use config format
- "advanced pyscript"  users can out out and have faster startup by just writing their config in JSON.


> **but** I've found the "official" [toml library](https://github.com/iarna/iarna-toml#tomlparsestr--object-example) which has a relatively small export and it produces a very nice error ... it adds, once minified and brotlied, just ~8K to the project

note that for pyscript classic we used fast-toml and toml-j0.4 (I think that in the latest incarnation, we were using @hoodmane's fork of toml-j0.4, I guess because he fixed some bugs), which I think they are even smaller.
According to my browser, fast-toml should be ~3.1 kb and toml-j0.4 ~6.6kb

**Comment by WebReflection** [comment]
I wouldn't mind using @hoodmane module but let's face reality: he's busy "*onboarding*" these days and if we find anything off I'd feel bad asking him to fix his own port of those libraries ... I don't know the state of those libraries neither:

  * are these abandoned so a fork was needed?
  * are these problematic still today, so that we need a fork?

> fast-toml should be ~3.1 kb

do we have any reason then to not use fast-toml as dependency?

**Comment by WebReflection** [comment]
OK, I had a look ... [fast-toml](https://www.npmjs.com/package/fast-toml) has poor adoption ... which means nothing, but the fact it's from 4 years ago @ TOML 0.5 rings a bell ...

> If we decide to support a half-baked custom format which resembles TOML, please don't call it TOML, else it just add confusion

how is serving for something called *Next* a 4yo version of an outdated standard, as the other project I've mentioned is at TOML v 1.something ???

Then again, we have 8 fields or less (or not much more) and I think those fields syntax is solid ... so if current TOML accept the syntax we described and there's no modern TOML syntax to represent the fields we're using, I think *fast-toml* it is.

**Comment by hoodmane** [comment]
I just changed package.json to add the type definitions to the file list. The change has been upstreamed and released in toml-j v1.1.2

**Comment by WebReflection** [comment]
@hoodmane awesome! I would use toml-j as it looks maintained and updated (8 months ago is way better than 4 years ago without any activity in the repo) but then I see LGPL-3.0 License ... do we have any concern around it? If not, I'll try this module instead and see how well it works.

**Comment by hoodmane** [comment]
Toml-j is mit licensed.

**Comment by WebReflection** [comment]
@hoodmane it doesn't exist in *npm* ... only result is this https://www.npmjs.com/search?q=toml-j and it's LGPL and it's also 16.5K https://cdn.jsdelivr.net/npm/@ltd/j-toml/+esm

**Comment by hoodmane** [comment]
We're talking about this one:
https://www.npmjs.com/package/toml-j0.4

**Comment by hoodmane** [comment]
Hmmm but there's a GitHub tag for 1.1.2 but it's not pushed to npm.

**Comment by WebReflection** [comment]
This file seems to be the latest https://cdn.jsdelivr.net/gh/jakwings/toml-j0.4/dist/toml-browser.js

**Comment by WebReflection** [comment]
FYI my latest MR uses LGPL 3.0 module via https://cdn.jsdelivr.net/npm/@ltd/j-toml/+esm ... I really would like to do something better than this as this module is not lightweight (dunno about fast) and the license is not MIT compatible ... are we stuck in here?

**Comment by hoodmane** [comment]
Maybe we can help fix the problems with mit-licensed toml-j0.4? Since it's half the size and generally works. It seems like the maintainer is open to accepting contributions.

I think LGPL might be troublesome for the downstream:

> if it is a "work that uses the library", then it must be possible for the software to be linked with a newer version of the LGPL-covered program. The most commonly used method for doing so is to use "a suitable [shared library](https://en.wikipedia.org/wiki/Library_(computing)#Shared_libraries) mechanism for linking". Alternatively, a [statically linked library](https://en.wikipedia.org/wiki/Statically_linked_library) is allowed if either source code or linkable object files are provided.

This could cause trouble for downstream projects if they want to follow the letter of the law since if they use PyScript in anything they need to provide a way to substitute a custom PyScript to follow the rules for the one LGPL piece, so it has the effect as if all of PyScript was LGPL even though only one piece of it is LGPL. This is unlike the MPL by the way. It's worse than MPL for compatibility since MPL merely requires that if you change an MPL'd source file the changes need to be public, but it doesn't require that it be possible to substitute a newer version of the library in a downstream application or that the process by which it is linked together be open.

**Comment by WebReflection** [comment]
OK, reading this issue I am very concerned about that module https://github.com/jakwings/toml-j0.4/issues/9

  * it's not maintained anymore (apparently)
  * the user lost the npm account meaning we cannot trust npm anumore (why don't people use 2FA?)
  * either somebody takes over that module if we think we really want to use it or we use that git repo with pinned version and we embed or build that as lazy artifact within out project ... this works for me but the first point of this list bothers me

**Comment by WebReflection** [comment]
So I went ahead and re-published that module in a CDN compliant way: https://github.com/jakwings/toml-j0.4/issues/9#issuecomment-1720963049 ... I will change the previous one to use this one instead.

**Comment by WebReflection** [comment]
MR landed.
